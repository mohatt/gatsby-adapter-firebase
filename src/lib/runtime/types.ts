import type { onRequest, HttpsOptions } from 'firebase-functions/v2/https'

export type FunctionHandler = Parameters<typeof onRequest>[0]
export type Request = Parameters<FunctionHandler>[0]
export type Response = Parameters<FunctionHandler>[1]

export type FunctionConfig = HttpsOptions
export type FunctionModule<
  THandler extends FunctionHandler = FunctionHandler,
  TConfig extends FunctionConfig = FunctionConfig,
> = THandler | { default: THandler; config?: TConfig }
