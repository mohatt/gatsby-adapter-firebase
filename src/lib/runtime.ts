export * from './runtime/helpers.js'
export * from './runtime/types.js'
