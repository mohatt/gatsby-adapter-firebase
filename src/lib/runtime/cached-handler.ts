import { initializeApp } from 'firebase-admin/app'
import { getStorage } from 'firebase-admin/storage'
import type { Bucket, File } from '@google-cloud/storage'
import type { OutgoingHttpHeader } from 'node:http'
import type { FunctionHandler, Request, Response } from './types.js'

export interface CachedResponseMetadata {
  status: number
  headers: Array<{ name: string; value: OutgoingHttpHeader }>
}

type AllowedMethod = 'GET' | 'HEAD'
const ALLOWED_METHODS: readonly AllowedMethod[] = ['GET', 'HEAD']
const HTTP_STATUS_METHOD_NOT_ALLOWED = 405

const CACHE_CONTROL_CACHEABLE = 'public, max-age=0, must-revalidate'
const CACHE_CONTROL_UNCACHEABLE = 'no-store'
const CACHE_METADATA_HEADER = 'X-Gatsby-Firebase-Cache'
const CACHE_HIT_VALUE = 'HIT'
const CACHE_MISS_VALUE = 'MISS'
const CACHE_PASS_VALUE = 'PASS'

// Hop-by-hop headers are not intended for cached payloads; filter them out.
const EXCLUDED_CACHE_HEADER_NAMES = [
  'connection',
  'keep-alive',
  'proxy-authenticate',
  'proxy-authorization',
  'te',
  'trailer',
  'transfer-encoding',
  'upgrade',
  'set-cookie', // prevent cross-user cookie replay
  'content-length', // derived from body
  'x-gatsby-firebase-cache', // internal metadata
]

// used for storage object naming and error reporting
const PREFIX = 'gatsby-adapter-firebase'

let cachedBucket: Bucket | null | undefined

// lazily resolve and memoize the default storage bucket; cache null if initialization fails
const getBucket = () => {
  if (cachedBucket !== undefined) return cachedBucket
  try {
    const app = initializeApp()
    cachedBucket = getStorage(app).bucket() as unknown as Bucket
  } catch (error) {
    cachedBucket = null
    console.error(`[gatsby-adapter-firebase] Failed to initialize Firebase Storage:`, error)
  }
  return cachedBucket
}

// normalize the request path so equivalent URLs land on the same cache entry
const normalizePath = (value: string | undefined) => {
  if (!value || value === '/') return '/'
  // strip query and hash
  const index = value.search(/[?#]/)
  const normalized = index === -1 ? value : value.slice(0, index)
  // ensure one leading slash
  return normalized.startsWith('/') ? normalized : `/${normalized}`
}

const prepareRequest = (originalReq: Request) => {
  const req = Object.create(originalReq) as Request

  const assignIfString = (key: string) => {
    if (typeof req[key] === 'string') {
      req[key] = normalizePath(req[key])
    }
  }

  assignIfString('url')
  assignIfString('originalUrl')
  req.query = Object.create(null)
  void req.path // trigger getter for reparse
  return req
}

// convert any express chunk shape into a Buffer so we can concatenate responses safely
const toBuffer = (chunk: unknown, encoding?: unknown): Buffer | null => {
  if (chunk == null) return null
  if (Buffer.isBuffer(chunk)) return chunk
  if (typeof chunk === 'string') {
    return Buffer.from(
      chunk,
      typeof encoding === 'string' ? (encoding as BufferEncoding) : undefined,
    )
  }
  if (chunk instanceof Uint8Array) return Buffer.from(chunk)
  return null
}

// capture cacheable headers (without hop-by-hop metadata) for storage
const createCachedHeaders = (res: Response, contentLength: number) => {
  const headers = Object.entries(res.getHeaders())
    .filter(([name, value]) => {
      if (!name || value == null) return false
      return !EXCLUDED_CACHE_HEADER_NAMES.includes(name.toLowerCase())
    })
    .map(([name, value]) => ({ name, value }))
  headers.push({ name: 'content-length', value: contentLength })
  headers.sort((a, b) => a.name.localeCompare(b.name))
  return headers
}

// encode function id + normalized path into a stable bucket object key
const createCacheKey = (functionId: string, req: Request) => {
  const path =
    (typeof req.originalUrl === 'string' && req.originalUrl) ||
    (typeof req.path === 'string' && req.path) ||
    (typeof req.url === 'string' && req.url) ||
    '/'
  const encoded = Buffer.from(path).toString('base64url')
  return `${PREFIX}/${functionId}/${encoded}.bin`
}

interface CachedResponse {
  metadata: CachedResponseMetadata
  body: Buffer
}

const readCachedResponse = async (file: File): Promise<CachedResponse | null> => {
  try {
    const [exists] = await file.exists()
    if (!exists) return null
    const [downloaded] = await file.download({ validation: false })
    const newlineIndex = downloaded.indexOf(0x0a) // '\n'
    if (newlineIndex === -1) {
      throw new Error(`Missing metadata delimiter`)
    }
    const metadata: CachedResponseMetadata = JSON.parse(
      downloaded.subarray(0, newlineIndex).toString('utf8'),
    )
    if (!metadata || typeof metadata.status !== 'number' || !Array.isArray(metadata.headers)) {
      throw new Error(`Invalid metadata`)
    }
    const body = downloaded.subarray(newlineIndex + 1)
    return { metadata, body }
  } catch (error) {
    console.error(`[${PREFIX}] Failed to read cached response for ${file.name}:`, error)
  }
  return null
}

// persist the captured payload for future hits
const writeCachedResponse = async (file: File, metadata: CachedResponseMetadata, body: Buffer) => {
  try {
    const header = `${JSON.stringify(metadata)}\n`
    const data = Buffer.concat([Buffer.from(header, 'utf8'), body])
    await file.save(data, {
      resumable: false,
      metadata: {
        contentType: 'application/octet-stream',
        metadata: { headerLength: header.length },
      },
    })
  } catch (error) {
    console.error(`[${PREFIX}] Failed to write cached response for ${file.name}:`, error)
  }
}

// cache successful/redirect/404 responses; everything else must re-run
const isCacheableStatus = (statusCode: number) =>
  (statusCode >= 200 && statusCode < 300) ||
  statusCode === 301 ||
  statusCode === 302 ||
  statusCode === 307 ||
  statusCode === 308 ||
  statusCode === 404

// Wrap the original handler with Firebase Storage backed response caching (2xx, 3xx, 404 only).
export const createCachedHandler = (handler: FunctionHandler, id: string): FunctionHandler => {
  return async (originalReq, res) => {
    const method = (originalReq.method?.toUpperCase() as AllowedMethod) ?? 'GET'
    if (!ALLOWED_METHODS.includes(method)) {
      res.status(HTTP_STATUS_METHOD_NOT_ALLOWED)
      res.set('allow', ALLOWED_METHODS.join(', '))
      res.set('cache-control', 'no-store')
      res.send('Method Not Allowed')
      return
    }

    const req = prepareRequest(originalReq)
    const bucket = getBucket()
    if (!bucket) {
      res.set(CACHE_METADATA_HEADER, CACHE_PASS_VALUE)
      if (!res.hasHeader('cache-control')) {
        res.set('cache-control', CACHE_CONTROL_UNCACHEABLE)
      }
      await handler(req, res)
      return
    }

    const cacheKey = createCacheKey(id, req)
    const file = bucket.file(cacheKey)

    const cached = await readCachedResponse(file)
    if (cached) {
      // apply cached headers to the response
      for (const header of cached.metadata.headers) {
        // we just pass the value we got from the cache, express should handle the rest
        res.set(header.name, header.value as any)
      }
      res.set(CACHE_METADATA_HEADER, CACHE_HIT_VALUE)
      res.status(cached.metadata.status)

      if (method === 'HEAD') {
        res.end()
        return
      }

      res.end(cached.body)
      return
    }

    let pendingWrites = 0
    let totalLength = 0
    const bufferList: Buffer[] = []
    // skip buffering HEAD requests
    // but we still have to run the same logic for both GET and HEAD for perfect header parity
    const shouldBuffer = method === 'GET'

    const addChunk = (chunk: unknown, encoding?: unknown) => {
      pendingWrites--
      if (shouldBuffer) {
        const buffer = toBuffer(chunk, encoding)
        if (buffer) {
          bufferList.push(buffer)
          totalLength += buffer.length
        }
      }
    }

    // intercept writes to buffer the body and set cache-control once status code is known
    const buildChunkArgs = (chunk: unknown, a?: unknown, b?: unknown) => {
      pendingWrites++

      // node allows: write(chunk), write(chunk, cb), write(chunk, encoding), write(chunk, encoding, cb)
      const encoding = typeof a === 'string' ? (a as BufferEncoding) : undefined
      const cb = (typeof a === 'function' ? a : typeof b === 'function' ? b : undefined) as
        | ((err?: Error) => void)
        | undefined

      if (!res.headersSent) {
        res.set(CACHE_METADATA_HEADER, CACHE_MISS_VALUE)
        if (!res.hasHeader('cache-control')) {
          const statusCode = res.statusCode ?? 200
          res.set(
            'cache-control',
            isCacheableStatus(statusCode) ? CACHE_CONTROL_CACHEABLE : CACHE_CONTROL_UNCACHEABLE,
          )
        }
      }

      // build safe argument list
      const callArgs = encoding ? [chunk, encoding] : [chunk]
      callArgs.push((err?: Error) => {
        cb?.(err)
        if (err == null) addChunk(chunk, encoding)
      })

      return callArgs
    }

    const originalWrite = res.write
    const originalEnd = res.end
    res.write = new Proxy(originalWrite, {
      apply(target, thisArg, args: [unknown]) {
        return Reflect.apply(target, thisArg, buildChunkArgs(...args))
      },
    })
    res.end = new Proxy(originalEnd, {
      apply(target, thisArg, args: [unknown]) {
        return Reflect.apply(target, thisArg, buildChunkArgs(...args))
      },
    })

    let done = false
    let queued = false

    const onClose = () => {
      if (done) return
      done = true
      cleanup()

      if (res.errored || !queued || pendingWrites > 0 || !shouldBuffer) {
        // aborted, errored, incomplete or head request -> skip
        return
      }

      // skip uncacheable responses
      const statusCode = res.statusCode
      if (!isCacheableStatus(statusCode)) {
        return
      }

      const body = Buffer.concat(bufferList, totalLength)
      const metadata: CachedResponseMetadata = {
        status: statusCode,
        headers: createCachedHeaders(res, totalLength),
      }

      void writeCachedResponse(file, metadata, body)
    }

    const onFinish = () => {
      queued = true

      // some providers might not emit 'close' and buffer the response
      if (pendingWrites === 0) {
        onClose()
      }
    }

    const cleanup = () => {
      res.off('finish', onFinish)
      res.off('close', onClose)
      res.write = originalWrite
      res.end = originalEnd
    }

    res.once('finish', onFinish)
    res.once('close', onClose)

    try {
      await handler(req, res)
    } catch (error) {
      cleanup()
      throw error
    }
  }
}
