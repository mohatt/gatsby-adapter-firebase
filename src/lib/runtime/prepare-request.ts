import { parse as parseCookies } from 'cookie'
import type { Request } from './types.js'

const normalizePath = (value: string | undefined) => {
  if (!value || value === '/') return '/'
  // strip query and hash
  const index = value.search(/[?#]/)
  const normalized = index === -1 ? value : value.slice(0, index)
  // ensure one leading slash
  return normalized.startsWith('/') ? normalized : `/${normalized}`
}

export const prepareRequest = (originalReq: Request, resetQuery?: boolean) => {
  const req = Object.create(originalReq) as Request

  // Gatsby includes cookie middleware
  const cookies = req.headers.cookie
  if (cookies) {
    req.cookies = parseCookies(cookies)
  }

  if (!resetQuery) {
    return req
  }

  // strip query and hash from urls
  if (typeof req.url === 'string') req.url = normalizePath(req.url)
  if (typeof req.originalUrl === 'string') req.originalUrl = normalizePath(req.originalUrl)

  // reset `query` which was set by `req.path` getter
  req.query = Object.create(null)
  // trigger getter for reparse
  void req.path

  return req
}
