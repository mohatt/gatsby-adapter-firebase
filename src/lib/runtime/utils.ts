import { App, getApp, initializeApp } from 'firebase-admin/app'
import { parse as parseCookies } from 'cookie'
import type { Request } from './types.js'

let app: App | undefined

export const getDefaultFirebaseApp = () => {
  if (app) return app
  try {
    app = getApp()
  } catch {
    app = initializeApp()
  }
  return app
}

const normalizePath = (value: string | undefined) => {
  if (!value || value === '/') return '/'
  // strip query and hash
  const index = value.search(/[?#]/)
  const normalized = index === -1 ? value : value.slice(0, index)
  // ensure one leading slash
  return normalized.startsWith('/') ? normalized : `/${normalized}`
}

export const prepareRequest = (originalReq: Request, resetQuery?: boolean) => {
  const req = Object.create(originalReq) as Request

  // Gatsby includes cookie middleware
  const cookies = req.headers.cookie
  if (cookies) {
    req.cookies = parseCookies(cookies)
  }

  if (!resetQuery) {
    return req
  }

  // strip query and hash from urls
  if (typeof req.url === 'string') req.url = normalizePath(req.url)
  if (typeof req.originalUrl === 'string') req.originalUrl = normalizePath(req.originalUrl)

  // reset `query` which was set by `req.path` getter
  req.query = Object.create(null)
  // trigger getter for reparse
  void req.path

  return req
}
