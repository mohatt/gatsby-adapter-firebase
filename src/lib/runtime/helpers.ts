import { onRequest, HttpsFunction } from 'firebase-functions/v2/https'
import { createCachedHandler } from './cached-handler.js'
import type { FunctionModule, FunctionHandler, FunctionConfig } from './types.js'

const isPlainObject = (value: unknown): value is Record<string, unknown> =>
  Object.prototype.toString.call(value) === '[object Object]'

const mergeHandlerOptions = <T extends FunctionConfig | undefined>(
  ...sources: Array<T | undefined>
): T | undefined => {
  const filtered = sources.filter((candidate): candidate is T => isPlainObject(candidate))
  if (filtered.length === 0) return undefined
  return Object.assign({}, ...filtered)
}

const resolveFunctionExports = (module: FunctionModule, id: string) => {
  let handler: FunctionHandler
  let config: FunctionConfig | undefined
  if ('default' in module) {
    // export default () => {}
    // export const options = {...}
    handler = module.default
    config = module.config
  } else {
    // module.exports = () => {}
    handler = module
  }

  if (typeof handler !== 'function') {
    throw new Error(`[gatsby-adapter-firebase] Expected function export for ${id} to be callable`)
  }

  return [handler, config] as const
}

export const createHttpsFunction = (
  module: FunctionModule,
  id: string,
  baseConfig?: FunctionConfig,
): HttpsFunction => {
  const [handler, configExport] = resolveFunctionExports(module, id)
  const config = mergeHandlerOptions(baseConfig, configExport)
  if (config) {
    return onRequest(config, handler)
  }
  return onRequest(handler)
}

export const createCachedHttpsFunction = (
  module: FunctionModule,
  id: string,
  baseConfig?: FunctionConfig,
): HttpsFunction => {
  const [handler, configExport] = resolveFunctionExports(module, id)
  const cachedHandler = createCachedHandler(handler, id)
  const options = mergeHandlerOptions(baseConfig, configExport)
  if (options) {
    return onRequest(options, cachedHandler)
  }
  return onRequest(cachedHandler)
}
