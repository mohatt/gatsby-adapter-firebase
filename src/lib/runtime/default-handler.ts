import type { FunctionHandler, FunctionMetadata } from './types.js'
import { prepareRequest } from './utils.js'

export const createDefaultHandler = (
  handler: FunctionHandler,
  meta: Pick<FunctionMetadata, 'id'>,
): FunctionHandler => {
  return async (originalReq, res) => {
    /**
     * Gatsby SSR matches page routes using req.url.
     * If query params are added, we end up with a req.url like /foo?q=bar.
     * Gatsby doesn't match that to /foo route, so we end up with a 404.
     * Here, we reset query params from only the SSR function.
     */
    const resetQuery = meta.id === 'ssr-engine'

    const req = prepareRequest(originalReq, resetQuery)

    await handler(req, res)
  }
}
