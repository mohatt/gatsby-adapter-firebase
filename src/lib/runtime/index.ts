import { onRequest, HttpsFunction } from 'firebase-functions/v2/https'
import { createDefaultHandler } from './default-handler.js'
import { createCachedHandler } from './cached-handler.js'
import type { FunctionModule, FunctionHandler, FunctionConfig, FunctionMetadata } from './types.js'

const resolveFunctionExports = (
  module: FunctionModule,
  meta: FunctionMetadata,
  baseConfig?: FunctionConfig,
) => {
  let gatsbyHandler: FunctionHandler
  let functionConfig: FunctionConfig | undefined

  if ('default' in module) {
    // export default () => {}
    // export const config = { firebase: {...} }
    gatsbyHandler = module.default
    functionConfig = module.config?.firebase
  } else {
    // module.exports = () => {}
    gatsbyHandler = module
  }

  if (typeof gatsbyHandler !== 'function') {
    throw new TypeError(
      `[gatsby-adapter-firebase] Expected function export for ${meta.id} to be callable`,
    )
  }

  const config: FunctionConfig = {
    invoker: 'public',
    ...baseConfig,
    ...functionConfig,
    labels: {
      ...baseConfig?.labels,
      ...functionConfig?.labels,
      generator: meta.generator,
    },
  }

  return [gatsbyHandler, config] as const
}

export const createHttpsFunction = (
  module: FunctionModule,
  meta: FunctionMetadata,
  baseConfig?: FunctionConfig,
): HttpsFunction => {
  const [handler, config] = resolveFunctionExports(module, meta, baseConfig)
  return onRequest(config, createDefaultHandler(handler, meta))
}

export const createCachedHttpsFunction = (
  module: FunctionModule,
  meta: FunctionMetadata,
  baseConfig?: FunctionConfig,
): HttpsFunction => {
  const [handler, config] = resolveFunctionExports(module, meta, baseConfig)
  return onRequest(config, createCachedHandler(handler, meta))
}
