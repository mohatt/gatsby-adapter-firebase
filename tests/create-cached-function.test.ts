import _ from 'lodash'
import express from 'express'
import request from 'supertest'
import { createCachedFunction } from '../src/lib/runtime.js'
import type { FunctionHandler } from '../src/lib/runtime/types.js'

const cache = new Map<string, { file: Buffer; metadata: object }>()

vi.mock('firebase-admin/app', () => ({
  initializeApp: vi.fn(() => ({})),
}))

vi.mock('firebase-admin/storage', () => ({
  getStorage: vi.fn(() => ({
    bucket: () => ({
      exists: vi.fn().mockResolvedValue([true]),
      file: (name: string) => ({
        async exists(): Promise<[boolean]> {
          return [cache.has(name)]
        },
        async download(): Promise<[Buffer]> {
          if (!cache.has(name)) {
            throw Object.assign(new Error('Not Found'), { code: 404 })
          }
          const payload = cache.get(name)
          if (!payload) {
            throw new Error(`No stored payload for ${name}`)
          }
          return [Buffer.from(payload.file)]
        },
        async save(data: string | Buffer, options?: { metadata?: object }) {
          const file = typeof data === 'string' ? Buffer.from(data, 'utf8') : Buffer.from(data)
          const { metadata } = options ?? {}
          cache.set(name, { file, metadata })
        },
      }),
    }),
  })),
}))

const createTestApp = (id: string, handler: FunctionHandler) => {
  const mockHandler = vi.fn(handler)
  const mockFunction = vi.fn(
    createCachedFunction(mockHandler, {
      id,
      name: `${id}-fn`,
      version: 'test-version',
      generator: 'test',
    }),
  )
  const server = express().disable('x-powered-by')
  server.all('*', (req, res) => {
    Promise.resolve(mockFunction(req as any, res)).catch((error) => {
      res.statusCode = 500
      res.end(error.message)
    })
  })
  return { handler: mockHandler, fn: mockFunction, agent: request(server) } as const
}

const assertResponse = (
  res: request.Test | request.Response,
  key: string,
): void | Promise<void> => {
  if ('then' in res) return res.then((r) => assertResponse(r, key))
  const snap = _.pick(res, ['status', 'headers', 'text', 'body', 'type'])
  if (snap.headers.date) snap.headers = { ...snap.headers, date: '<any>' }
  expect(snap).toMatchSnapshot(`response[${key}]`)
}

const assertCacheState = (encoding: BufferEncoding = 'utf8') => {
  const decodedCache = new Map(
    [...cache.entries()].map(([key, entry]) => {
      const { file, metadata } = entry
      const newlineIndex = file.indexOf(0x0a)
      const header = JSON.parse(file.subarray(0, newlineIndex).toString('utf8'))
      const body = file.subarray(newlineIndex + 1).toString(encoding)
      return [key, { file: { header, body }, metadata }]
    }),
  )
  expect(decodedCache).toMatchSnapshot('cache')
}

describe('createCachedFunction()', () => {
  beforeEach(() => {
    cache.clear()
    vi.clearAllMocks()
  })

  it('serves streaming responses correctly', async () => {
    const { handler, agent } = createTestApp('function-id', async (_req, res) => {
      res.statusCode = 200
      res.setHeader('Test-Header', 'test-value')
      res.write('hello')
      res.write(', ')
      res.end('world')
    })
    await assertResponse(agent.get('/hello?query=ignore'), 'miss')
    await assertResponse(agent.get('/hello?param=ignored'), 'hit')
    expect(handler).toHaveBeenCalledTimes(1)
    assertCacheState()
  })

  it('does not cache non-cacheable status codes', async () => {
    const { handler, agent } = createTestApp('non-cacheable', async (_req, res) => {
      res.statusCode = 500
      res.end('error')
    })
    await assertResponse(agent.get('/error'), 'miss')
    await assertResponse(agent.get('/error'), 'repeat')
    expect(handler).toHaveBeenCalledTimes(2)
    assertCacheState()
  })

  it('serves cached GET bodies to HEAD requests', async () => {
    const { handler, agent } = createTestApp('head', async (_req, res) => {
      res.statusCode = 200
      res.end('payload')
    })
    await assertResponse(agent.get('/head'), 'get')
    await assertResponse(agent.head('/head'), 'head')
    expect(handler).toHaveBeenCalledTimes(1)
    assertCacheState()
  })

  it('strips query strings before invoking the wrapped handler', async () => {
    const { handler, agent } = createTestApp('strip-query', async (_req, res) => {
      res.status(204).end()
    })

    await agent.get('/stripped/path?foo=bar&baz=qux')
    const [req] = handler.mock.lastCall ?? []
    expect(req).toBeDefined()
    expect(req.url).toBe('/stripped/path')
    expect(req.originalUrl).toBe('/stripped/path')
    expect(req.query).toEqual({})
    expect((req as any)._parsedUrl).toMatchObject({
      search: null,
      query: null,
      path: '/stripped/path',
    })
  })

  it('caches 404 responses', async () => {
    const { handler, agent } = createTestApp('not-found', async (_req, res) => {
      res.statusCode = 404
      res.end('missing')
    })
    await assertResponse(agent.get('/missing'), 'miss')
    await assertResponse(agent.get('/missing'), 'hit')
    expect(handler).toHaveBeenCalledTimes(1)
    assertCacheState()
  })

  it('handles write errors correctly', async () => {
    const onError = vi.fn()
    const { handler, agent } = createTestApp('error', async (_req, res) => {
      res.once('error', onError)
      res.write('1-', () => res.write('2-'))
      res.end('3')
    })
    await assertResponse(agent.get('/error'), 'miss')
    await assertResponse(agent.get('/error'), 'repeat')
    expect(handler).toHaveBeenCalledTimes(2)
    expect(onError).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        message: 'write after end',
        code: 'ERR_STREAM_WRITE_AFTER_END',
      }),
    )
    assertCacheState()
  })

  it('skips stale cache entries', async () => {
    const { handler, agent } = createTestApp('version-check', async (_req, res) => {
      res.statusCode = 200
      res.setHeader('test-header', 'fresh-value')
      res.end('fresh-response')
    })

    const encodedPath = Buffer.from('/fresh').toString('base64url')
    const key = `.gatsby-adapter-firebase/version-check/${encodedPath}.bin`
    const staleHeader = JSON.stringify({
      status: 200,
      headers: [
        { name: 'cache-control', value: 'public, max-age=0, must-revalidate' },
        { name: 'content-length', value: 5 },
      ],
      version: 'old-version',
    })

    cache.set(key, {
      file: Buffer.concat([Buffer.from(`${staleHeader}\n`, 'utf8'), Buffer.from('stale', 'utf8')]),
      metadata: { metadata: { headerLength: staleHeader.length + 1 } },
    })

    await assertResponse(agent.get('/fresh'), 'miss')
    await assertResponse(agent.get('/fresh'), 'hit')
    expect(handler).toHaveBeenCalledTimes(1)
    assertCacheState()
  })
})
