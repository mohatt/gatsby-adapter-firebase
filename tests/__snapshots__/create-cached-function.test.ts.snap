// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createCachedFunction() > caches 404 responses > cache 1`] = `
Map {
  ".gatsby-adapter-firebase/not-found/L21pc3Npbmc.bin" => {
    "file": {
      "body": "missing",
      "header": {
        "headers": [
          {
            "name": "cache-control",
            "value": "public, max-age=0, must-revalidate",
          },
          {
            "name": "content-length",
            "value": 7,
          },
        ],
        "status": 404,
        "version": "test-version",
      },
    },
    "metadata": {
      "contentType": "application/octet-stream",
      "metadata": {
        "headerLength": 158,
      },
    },
  },
}
`;

exports[`createCachedFunction() > caches 404 responses > response[hit] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "7",
    "date": "<any>",
    "x-gatsby-firebase-cache": "HIT",
  },
  "status": 404,
  "text": "missing",
  "type": "",
}
`;

exports[`createCachedFunction() > caches 404 responses > response[miss] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "7",
    "date": "<any>",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 404,
  "text": "missing",
  "type": "",
}
`;

exports[`createCachedFunction() > does not cache non-cacheable status codes > cache 1`] = `Map {}`;

exports[`createCachedFunction() > does not cache non-cacheable status codes > response[miss] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "no-store",
    "connection": "close",
    "content-length": "5",
    "date": "<any>",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 500,
  "text": "error",
  "type": "",
}
`;

exports[`createCachedFunction() > does not cache non-cacheable status codes > response[repeat] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "no-store",
    "connection": "close",
    "content-length": "5",
    "date": "<any>",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 500,
  "text": "error",
  "type": "",
}
`;

exports[`createCachedFunction() > handles write errors correctly > cache 1`] = `Map {}`;

exports[`createCachedFunction() > handles write errors correctly > response[miss] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "date": "<any>",
    "transfer-encoding": "chunked",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 200,
  "text": "1-3",
  "type": "",
}
`;

exports[`createCachedFunction() > handles write errors correctly > response[repeat] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "date": "<any>",
    "transfer-encoding": "chunked",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 200,
  "text": "1-3",
  "type": "",
}
`;

exports[`createCachedFunction() > serves cached GET bodies to HEAD requests > cache 1`] = `
Map {
  ".gatsby-adapter-firebase/head/L2hlYWQ.bin" => {
    "file": {
      "body": "payload",
      "header": {
        "headers": [
          {
            "name": "cache-control",
            "value": "public, max-age=0, must-revalidate",
          },
          {
            "name": "content-length",
            "value": 7,
          },
        ],
        "status": 200,
        "version": "test-version",
      },
    },
    "metadata": {
      "contentType": "application/octet-stream",
      "metadata": {
        "headerLength": 158,
      },
    },
  },
}
`;

exports[`createCachedFunction() > serves cached GET bodies to HEAD requests > response[get] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "7",
    "date": "<any>",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 200,
  "text": "payload",
  "type": "",
}
`;

exports[`createCachedFunction() > serves cached GET bodies to HEAD requests > response[head] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "7",
    "date": "<any>",
    "x-gatsby-firebase-cache": "HIT",
  },
  "status": 200,
  "text": undefined,
  "type": "",
}
`;

exports[`createCachedFunction() > serves streaming responses correctly > cache 1`] = `
Map {
  ".gatsby-adapter-firebase/function-id/L2hlbGxv.bin" => {
    "file": {
      "body": "hello, world",
      "header": {
        "headers": [
          {
            "name": "cache-control",
            "value": "public, max-age=0, must-revalidate",
          },
          {
            "name": "content-length",
            "value": 12,
          },
          {
            "name": "test-header",
            "value": "test-value",
          },
        ],
        "status": 200,
        "version": "test-version",
      },
    },
    "metadata": {
      "contentType": "application/octet-stream",
      "metadata": {
        "headerLength": 203,
      },
    },
  },
}
`;

exports[`createCachedFunction() > serves streaming responses correctly > response[hit] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "12",
    "date": "<any>",
    "test-header": "test-value",
    "x-gatsby-firebase-cache": "HIT",
  },
  "status": 200,
  "text": "hello, world",
  "type": "",
}
`;

exports[`createCachedFunction() > serves streaming responses correctly > response[miss] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "date": "<any>",
    "test-header": "test-value",
    "transfer-encoding": "chunked",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 200,
  "text": "hello, world",
  "type": "",
}
`;

exports[`createCachedFunction() > skips stale cache entries > cache 1`] = `
Map {
  ".gatsby-adapter-firebase/version-check/L2ZyZXNo.bin" => {
    "file": {
      "body": "fresh-response",
      "header": {
        "headers": [
          {
            "name": "cache-control",
            "value": "public, max-age=0, must-revalidate",
          },
          {
            "name": "content-length",
            "value": 14,
          },
          {
            "name": "test-header",
            "value": "fresh-value",
          },
        ],
        "status": 200,
        "version": "test-version",
      },
    },
    "metadata": {
      "contentType": "application/octet-stream",
      "metadata": {
        "headerLength": 204,
      },
    },
  },
}
`;

exports[`createCachedFunction() > skips stale cache entries > response[hit] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "14",
    "date": "<any>",
    "test-header": "fresh-value",
    "x-gatsby-firebase-cache": "HIT",
  },
  "status": 200,
  "text": "fresh-response",
  "type": "",
}
`;

exports[`createCachedFunction() > skips stale cache entries > response[miss] 1`] = `
{
  "body": {},
  "headers": {
    "cache-control": "public, max-age=0, must-revalidate",
    "connection": "close",
    "content-length": "14",
    "date": "<any>",
    "test-header": "fresh-value",
    "x-gatsby-firebase-cache": "MISS",
  },
  "status": 200,
  "text": "fresh-response",
  "type": "",
}
`;
